### PR TITLE
Migrate Expr and Lexer to PHP 8 syntax

### PR DESCRIPTION
**Pull request by greg0ire**

### CODE REVIEW

File: lib/Doctrine/ORM/Query/Expr/Comparison.php [hunk 1]
```diff
@@ -4,51 +4,47 @@
...
 
     /**
      * Creates a comparison expression with the given arguments.
      *
-     * @param mixed  $leftExpr
      * @param string $operator
-     * @param mixed  $rightExpr
      */
-    public function __construct(protected $leftExpr, protected $operator, protected $rightExpr)
+    public function __construct(protected mixed $leftExpr, protected $operator, protected mixed $rightExpr)
```
Comment (SenseException) [inline lib/Doctrine/ORM/Query/Expr/Comparison.php]: If `$operator` is string, why no type hint here?

File: lib/Doctrine/ORM/Query/Expr/Composite.php [hunk 2]
```diff
@@ -31,8 +30,7 @@ public function __toString()
         return implode($this->separator, $components);
     }
 
-    /** @param string|object $part */
-    private function processQueryPart($part): string
+    private function processQueryPart(string|object $part): string
```
Comment (SenseException) [inline lib/Doctrine/ORM/Query/Expr/Composite.php]: I wonder if `string|Stringable` would be more fitting here.

File: lib/Doctrine/ORM/Query/Expr/From.php [hunk 3]
```diff
@@ -4,42 +4,44 @@
...
  *
  * @link    www.doctrine-project.org
  */
-class From
+class From implements Stringable
 {
     /**
      * @param class-string $from    The class name.
      * @param string       $alias   The alias of the class.
      * @param string       $indexBy The index for the from.
```
Comment (SenseException) [inline lib/Doctrine/ORM/Query/Expr/From.php]: Can be removed now.

File: lib/Doctrine/ORM/Query/Expr/Select.php [hunk 4]
```diff
@@ -11,20 +11,17 @@
...
-    protected $preSeparator = '';
-
-    /** @var string */
-    protected $postSeparator = '';
+    protected string $preSeparator  = '';
+    protected string $postSeparator = '';
 
     /** @var string[] */
-    protected $allowedClasses = [Func::class];
+    protected array $allowedClasses = [Func::class];
```
Comment (SenseException) [inline lib/Doctrine/ORM/Query/Expr/Select.php:18]: I don't know if that belongs into the scope (or if it fits here): `class-string[]` would be a nice addition.

**Review by SenseException — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Query/Expr/Select.php (hunk 4 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query/Expr/Select.php:18]: That's something that should be done on 2.14.x

Comment (greg0ire) [reply lib/Doctrine/ORM/Query/Expr/Select.php:18]: And IIRC, it's a rabbit hole if you want to specify something inside the `<>` of `class-string<>` :stuck_out_tongue:

File: lib/Doctrine/ORM/Query/Expr/Comparison.php (hunk 1 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query/Expr/Comparison.php]: Not sure why I left that one out, let's use a type hint.

File: lib/Doctrine/ORM/Query/Expr/Select.php (hunk 4 above)
Comment (SenseException) [reply lib/Doctrine/ORM/Query/Expr/Select.php:18]: Let's revisit this in 2.14 again.

**Review by SenseException — APPROVED** [review]

**Pull request merged**